metric/metric.go

```diff
@@ -8,20 +8,66 @@ import (
 	"github.com/prometheus/client_golang/prometheus/promauto"
 )
 
+const namespace = "gatus" // The prefix of the metrics
+
 var (
-	// This will be initialized once PublishMetricsForEndpoint.
-	// The reason why we're doing this is that if metrics are disabled, we don't want to initialize it unnecessarily.
-	resultCount *prometheus.CounterVec = nil
+	initializedMetrics bool // Whether the metrics have been initialized
+
+	resultTotal                        *prometheus.CounterVec
+	resultDurationSeconds              *prometheus.GaugeVec
+	resultConnectedTotal               *prometheus.CounterVec
+	resultCodeTotal                    *prometheus.CounterVec
+	resultCertificateExpirationSeconds *prometheus.GaugeVec
 )
 
+func initializePrometheusMetrics() {
+	resultTotal = promauto.NewCounterVec(prometheus.CounterOpts{
+		Namespace: namespace,
+		Name:      "results_total",
+		Help:      "Number of results per endpoint",
+	}, []string{"key", "group", "name", "type", "success"})
+	resultDurationSeconds = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: namespace,
+		Name:      "results_duration_seconds",
+		Help:      "Duration of the request in seconds",
+	}, []string{"key", "group", "name", "type"})
+	resultConnectedTotal = promauto.NewCounterVec(prometheus.CounterOpts{
+		Namespace: namespace,
+		Name:      "results_connected_total",
+		Help:      "Total number of results in which a connection was successfully established",
+	}, []string{"key", "group", "name", "type"})
```

Comment:
I'm wondering how useful this is compared to just using `results_total`'s `success` field?

Let's say that you had this configuration:
```yaml
endpoints:
  - name: unreachable-ip
    url: "icmp://123.23.45.75"
    interval: 5m
    conditions:
      - "[CONNECTED] == false"
```
`success` would be `true` because the condition `[CONNECTED] == false` is true.

You could make a case that for `HTTP` endpoints, you'd have `success` as `false` even if you had the same condition:
```yaml
endpoints:
  - name: secret-host-that-should-not-be-reachable
    url: "http://top-secret-host-that-is-not-reachable/health"
    interval: 5m
    conditions:
      - "[CONNECTED] == false"
```
the reason being that there would be an error like `no such host`, which would make the result evaluate as a failure despite passing all conditions.

---

TL;DR: When would you use `results_connected_total` instead of `results_total`?

Reply:
for this case
```
- name: website
  url: "https://twin.sh/healthssss"
  interval: 5s
  conditions:
  - "[CONNECTED] == true"
  - "[STATUS] == 200"
```

assume I have a http check, but the target path return 404 code not 200.
And I can check ```results_connected_total``` increase for connection is still working, but ```results_total``` is not success for code != 200.